### PR TITLE
Add preliminary support for adding non-KCP managed environments to the AppStudio Environment API

### PR DESCRIPTION
**Pull request by jgwest**
#### Description:
- This PR allows the user to specify a `Secret`, APIURL, and target namespace, in the definition of the `Environment` CR. When these are specified, the Environment represents a target cluster to deploy to. These values describe the credentials required to connect to the cluster, and we use them to configure Argo CD to do the deployment.
- With this PR, now whenever a `ApplicationSnapshotEnvironmentBinding` is created, it will use the above values when generating the `GitOpsDeployment` resources for the component. 
- This causes Argo CD to deploy to this environment, using these credentials.
- More information on the all the Environment APIs is available from [here](https://docs.google.com/document/d/1-_rWLgALd5pdSlqNNcQ5FSrD00fZb0l_exU-_FiL68o/edit#)

### CODE REVIEW

File: appstudio-controller/controllers/appstudio.redhat.com/applicationsnapshotenvironmentbinding_controller.go [hunk 1]
```diff
@@ -165,11 +194,26 @@ func processExpectedGitOpsDeployment(ctx context.Context, expectedGitopsDeployme
...
 
-func generateExpectedGitOpsDeployment(component appstudioshared.ComponentStatus, binding appstudioshared.ApplicationSnapshotEnvironmentBinding) apibackend.GitOpsDeployment {
+// GenerateBindingGitOpsDeploymentName generates the name that will be used for a given GitOpsDeployment of a binding
+func GenerateBindingGitOpsDeploymentName(binding appstudioshared.ApplicationSnapshotEnvironmentBinding, componentName string) string {
+
+	expectedName := binding.Name + "-" + binding.Spec.Application + "-" + binding.Spec.Environment + "-" + componentName
+
+	// If the length of the GitOpsDeployment exceeds the K8s maximum, shorten it to just binding+component
+	if len(expectedName) > 250 {
+		expectedName = binding.Name + "-" + componentName
```
Comment (drpaneas) [inline appstudio-controller/controllers/appstudio.redhat.com/applicationsnapshotenvironmentbinding_controller.go:204]: are we certain that `expectedName = binding.Name + "-" + componentName` is going to be **always** < = 250 length?

Consider shortening the string to be on the safe side: `return shorten(expectedName, 250)`

Feel free to grab my shorten function implementation: https://goplay.tools/snippet/v4X0lF-9z3G

File: appstudio-controller/controllers/appstudio.redhat.com/environment_controller.go [hunk 2]
```diff
@@ -39,21 +45,146 @@ type EnvironmentReconciler struct {
...
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      req.Name,
+			Namespace: req.Namespace,
+		},
+	}
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(environment), environment); err != nil {
+
+		if apierr.IsNotFound(err) {
+			// A) The Environment resource could not be found: the owner reference on the GitOpsDeploymentManagedEnvironment
+			// should ensure that it is cleaned up, so no more work is required.
```
Comment (drpaneas) [inline appstudio-controller/controllers/appstudio.redhat.com/environment_controller.go:67]: Do you think we need a `log.Info` here?

File: appstudio-controller/controllers/appstudio.redhat.com/environment_controller.go [hunk 3]
```diff
@@ -39,21 +45,146 @@ type EnvironmentReconciler struct {
...
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(&currentManagedEnv), &currentManagedEnv); err != nil {
+
+		if apierr.IsNotFound(err) {
+			// B) The GitOpsDeploymentManagedEnvironment doesn't exist, so needs to be created.
+
+			if err := r.Client.Create(ctx, desiredManagedEnv); err != nil {
+				return ctrl.Result{}, fmt.Errorf("unable to create new GitOpsDeploymentManagedEnvironment: %v", err)
+			}
+
+			// Success: the resource has been created.
```
Comment (drpaneas) [inline appstudio-controller/controllers/appstudio.redhat.com/environment_controller.go:92]: Same question here, do we need a `log.Info` informing that the object is created?

**Review by drpaneas — APPROVED** [review]
Great work here @jgwest , easy to understand. code, very helpful comments and sufficient testing. The PR was filling `golint` so I've fixed that already to be in a mergeable state. I've put some comments, in the review, feel free to take a look and address only those that make sense to you.

**Review by samyak-jn — APPROVED** [review]
/gtm, thanks @jgwest for mentioning comments before the implementations easy to navigate, and Panos did a good job with cosmetic changes suggestions.

My concerns is with the failing e2e tests, I can see there is a timeout error again, which I think we removed it when we merged my PR, do you think we missed something?

Ref from logs:
```
TEP: Delete old namespaces, and kube-system resources 07/14/22 07:31:20.414
  << End Captured GinkgoWriter Output
  Expected success, but got an error:
      <*errors.errorString | 0xc0004be5d0>: {
          s: "timed out waiting for the condition",
      }
      timed out waiting for the condition
  In [BeforeEach] at: /go/src/github.com/redhat-appstudio/managed-gitops/tests-e2e/core/argocd_instance_test.go:34
------------------------------
```

**Comment by drpaneas** [comment]
> /gtm, thanks @jgwest for mentioning comments before the implementations easy to navigate, and Panos did a good job with cosmetic changes suggestions.
> 
> My concerns is with the failing e2e tests, I can see there is a timeout error again, which I think we removed it when we merged my PR, do you think we missed something?
> 
> Ref from logs:
> 
> ```
> TEP: Delete old namespaces, and kube-system resources 07/14/22 07:31:20.414
>   << End Captured GinkgoWriter Output
>   Expected success, but got an error:
>       <*errors.errorString | 0xc0004be5d0>: {
>           s: "timed out waiting for the condition",
>       }
>       timed out waiting for the condition
>   In [BeforeEach] at: /go/src/github.com/redhat-appstudio/managed-gitops/tests-e2e/core/argocd_instance_test.go:34
> ------------------------------
> ```

The testsuite ended after 2h18m22s which sounds that it's not one of our timeout but a timeout from Prow. Let me test this locally ...

**Comment by samyak-jn** [comment]
> The PR seems to be breaking e2e tests. I've just tested it locally as well. The tests keep running for ever -- which is why Prow CI failed after ~2 hours.

Good to know, I figured out the same while looking for the logs!

**Comment by drpaneas** [comment]
> >

Not really, I've just run 2 times against master and I've got an error only the second time:

```
• [FAILED] [163.782 seconds]
Argo CD Application
/Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/tests-e2e/core/argocd_application_test.go:20
  Creating GitOpsDeployment should result in an Argo CD Application
  /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/tests-e2e/core/argocd_application_test.go:21
    [It] Argo CD Application should have has prune, allowEmpty and selfHeal enabled
    /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/tests-e2e/core/argocd_application_test.go:22

  Begin Captured GinkgoWriter Output >>
    STEP: create a new GitOpsDeployment CR 07/14/22 13:46:38.301
    STEP: GitOpsDeployment should have expected health and status 07/14/22 13:46:42.554
  << End Captured GinkgoWriter Output

  Timed out after 120.000s.
  Expected
      <bool>: false
  to be true
  In [It] at: /Users/drpaneas/gocode/src/github.com/redhat-appstudio/managed-gitops/tests-e2e/core/argocd_application_test.go:33
```

**Comment by jgwest** [comment]
@drpaneas @samyak-jn fixed the failing E2E test, and made a bunch of improvements to the `EnsureCleanSlate` function which should now do a more thorough job of cleaning up before each test.  Also made some improvements to the stability of a few other tests.

**Comment by drpaneas** [comment]
Great work Jonathan!

**Pull request merged**